app/hooks/use-event.ts

```diff
@@ -1,16 +1,22 @@
-import { useEffect } from "react";
-import type { SocketType } from "@/socket";
+import { useEffect, useState } from "react";
+import { socket } from "@/socket";
 import type { ServerToClientEvents } from "@/types";
 
-export const useEvent = (
-  socket: SocketType,
-  event: keyof ServerToClientEvents,
-  callback: (...args: any[]) => void,
-) => {
+export function useEvent<T extends keyof ServerToClientEvents>(event: T) {
+  const [result, setResult] = useState<
+    Parameters<ServerToClientEvents[T]> | []
+  >([]);
+
   useEffect(() => {
-    socket.on(event, callback);
+    const callback = (...args: Parameters<ServerToClientEvents[T]>) => {
+      setResult(args);
+    };
+
+    socket.on(event, callback as never);
     return () => {
-      socket.off(event, callback);
+      socket.off(event, callback as never);
```

Comment:
now that I think about it, maybe casting here is better if we need to refactor in the future (e.g. if socket.io fixes their damn types). I'd also make a comment about it for future reference
```suggestion
    const callback = ((...args: Parameters<ServerToClientEvents[T]>) => {
      setResult(args);
    }) as never; // unable to pass as a listener due to the socket.io types

    socket.on(event, callback);
    return () => {
      socket.off(event, callback);
```